src/mito2/src/schedule/scheduler.rs

```diff
@@ -0,0 +1,242 @@
+
+pub type Job = Pin<Box<dyn Future<Output = ()> + Send>>;
+
+///The state of scheduler
+const STATE_RUNNING: u8 = 0;
+const STATE_STOP: u8 = 1;
+const STATE_AWAIT_TERMINATION: u8 = 2;
+
+/// The consumer count
+const CONSUMER_NUM: u8 = 2;
+
+/// [Scheduler] defines a set of API to schedule Jobs
+#[async_trait::async_trait]
+pub trait Scheduler {
+    /// Schedules a Job
+    async fn schedule(&mut self, req: Job) -> Result<()>;
+
+    /// Stops scheduler. If `await_termination` is set to true, the scheduler will
+    /// wait until all tasks are processed.
+    async fn stop(&mut self, await_termination: bool) -> Result<()>;
+}
+
+/// Request scheduler based on local state.
+pub struct LocalScheduler {
+    /// Sends jobs to flume bounded channel
+    sender: Option<flume::Sender<Job>>,
+    /// Task handles
+    handles: Vec<Option<JoinHandle<()>>>,
+    /// Token used to halt the scheduler
+    cancel_token: CancellationToken,
+    /// State of scheduler
+    state: Arc<AtomicU8>,
+}
+
+impl LocalScheduler {
+    /// cap: flume bounded cap
+    /// num: the number of bounded receiver
+    pub fn new(cap: usize, num: usize) -> Self {
+        let (tx, rx) = flume::bounded(cap);
+        let token = CancellationToken::new();
+        let state = Arc::new(AtomicU8::new(STATE_RUNNING));
+
+        let mut handles = Vec::with_capacity(num);
+
+        for _ in 0..num {
+            let child = token.child_token().clone();
+            let receiver = rx.clone();
+            let state = Arc::clone(&state);
+            let handle = tokio::spawn(async move {
+                while state.load(Ordering::Relaxed) == STATE_RUNNING {
+                    info!("Task scheduler loop.");
+                    tokio::select! {
+                        _ = child.cancelled() => {
+                            info!("Task scheduler cancelled.");
+                            return;
+                        }
+                        req_opt = receiver.recv_async() =>{
+                            if let Ok(req) = req_opt{
+                                req.await;
+                            }
+                        }
+                    }
+                }
+                // For correctness, we need to handle all tasks
+                if state.load(Ordering::Relaxed) == STATE_AWAIT_TERMINATION {
+                    while let Ok(req) = receiver.recv() {
+                        req.await;
+                    }
+                }
+            });
+            handles.push(Some(handle));
+        }
+
+        Self {
+            sender: Some(tx),
+            cancel_token: token,
+            handles,
+            state,
+        }
+    }
+
+    #[inline]
+    fn running(&self) -> bool {
+        self.state.load(Ordering::Relaxed) == STATE_RUNNING
+    }
+}
+
+#[async_trait::async_trait]
+impl Scheduler for LocalScheduler {
+    async fn schedule(&mut self, req: Job) -> Result<()> {
+        self.sender.as_mut().unwrap().send_async(req).await.unwrap();
```

Comment:
Do not simply unwrap `self.sender` here since it would be a None if the scheduler's been stopped. Prefer:
```suggestion
        let Some(sender) = &self.sender else { return IllegalStateSnafu{}.fail() };
        sender.send(req).context(ScheduleRequestSnafu)
```

Reply:
Also, the sync version `flume::Sender::send` would be enough.